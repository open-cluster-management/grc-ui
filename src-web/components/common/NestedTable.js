/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableRowExpanded,
  TableData,
  TableHeader,
  DataTable
} from 'carbon-components-react'
import msgs from '../../../nls/platform.properties'
import resources from '../../../lib/shared/resources'

const { TableSelectRow, TableCell } = DataTable

resources(() => {
  require('../../../scss/nested-table.scss')
})

class NestedTable extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      toggle: [],
    }
  }

  toggleRow = index => () => {
    this.setState(preState => {
      const toggle = [...preState.toggle]
      toggle[index] = toggle[index] ? !toggle[index] : true
      return { toggle }
    })
  };

  getToggleState = index => this.state.toggle[index] ? this.state.toggle[index] : false

  getItemCheckedState = id => !!this.props.selectedItems.find(item => item.id === id)

  updateSelectedState = row => () => {
    const { selectedItems, selectionChanged } = this.props
    let newSelectedItems = [...selectedItems]
    if (selectedItems.find(item => item.id === row.id)) {
      const itemIndex = selectedItems.findIndex(item => item.id === row.id)
      newSelectedItems.splice(itemIndex, 1)
    } else {
      newSelectedItems = [...selectedItems, row]
    }
    selectionChanged && selectionChanged(newSelectedItems)
  }

  createTableSelectRow = (rows) => (
    <Table>
      <TableBody>
        {rows.map(row => (
          <TableRow key={row.id} even={false} >
            <TableSelectRow
              onSelect={this.updateSelectedState(row)}
              checked={this.getItemCheckedState(row.id)}
              ariaLabel={'tableSelectRow'}
              id={row.id}
              name={row.name}
            />
            <TableCell key={row.value}>{row.name}</TableCell>
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );


  render() {
    const {
      availableItems,
      header
    } = this.props

    const createExpandedRows = (item, index, selectionChanged) => {
      const toggleState = this.getToggleState(index)
      return (
        <TableRowExpanded expanded={toggleState} colSpan={4} key={`c${index}`}>
          {this.createTableSelectRow(item.value, selectionChanged)}
        </TableRowExpanded>
      )
    }

    const rowData = availableItems.map((item, index) => {
      const toggleState = this.getToggleState(index)
      return [
        <TableRow key={`b${item.key}`} even={false} className={'nested-table-select-row'}>
          <TableData
            onClick={this.toggleRow(index)}
            key={`a${item.key}`}
            expanded={toggleState}
            className={'nest-table-expand-icon'}
          />
          <TableData key={`d${item.key}`} className={'nest-table-expand-name'}>
            {msgs.get(`modal.formfield.${item.key}`, this.context.locale)}
          </TableData>
        </TableRow>,
        createExpandedRows(item, index)
      ]
    })

    return (
      <div className={'nested-table-main'}>
        <Table>
          <TableHead>
            <TableRow header className={'nested-table-header-row'}>
              <TableHeader />
              <TableHeader className={'nested-table-header'}>
                <p>{header}  </p>
                <p>({this.props.selectedItems.length} {msgs.get('nested.table.selected', this.context.locale)})</p>
              </TableHeader>
            </TableRow>
          </TableHead>
          <TableBody className={'nested-table-body'}>{rowData}</TableBody>
        </Table>
      </div>
    )
  }
}

NestedTable.propTypes = {
  availableItems: PropTypes.array,
  header: PropTypes.string,
  selectedItems: PropTypes.array,
  selectionChanged: PropTypes.func
}

export default NestedTable
