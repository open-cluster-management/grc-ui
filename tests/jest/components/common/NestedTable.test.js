/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { shallow } from 'enzyme'

import NestedTable from '../../../../src-web/components/common/NestedTable'

describe('NestedTable component 1', () => {
  const fn = jest.fn()
  const tagFilterProps = {
    selectedItems: [
      {
        'name': 'cloud=IBM',
        'id': 'cloud=IBM',
        'key': 'cloud',
        'value': 'IBM',
        'type': 'clusterLabel'
      },
    ],
    availableItems: [
      {
        key: 'clusterLabel',
        value: [
          {
            'name': 'cloud=IBM',
            'id': 'cloud=IBM',
            'key': 'cloud',
            'value': 'IBM',
            'type': 'clusterLabel'
          },
          {
            'name': 'clusterip=9.42.23.230',
            'id': 'clusterip=9.42.23.230',
            'key': 'clusterip',
            'value': '9.42.23.230',
            'type': 'clusterLabel'
          }
        ]
      }
    ]
  }
  it('renders as expected', () => {
    const component = shallow(
      <NestedTable  {...tagFilterProps} header={'header'} selectionChanged={fn} />
    )
    expect(component).toMatchSnapshot()
    component.find('.nest-table-expand-icon').at(0).simulate('click')
    expect(component).toMatchSnapshot()
  })
})
