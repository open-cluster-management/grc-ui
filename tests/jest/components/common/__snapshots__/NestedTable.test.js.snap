// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedTable component 1 renders as expected 1`] = `ShallowWrapper {}`;

exports[`NestedTable component 1 renders as expected 2`] = `ShallowWrapper {}`;
